ccp/lib/clients/openshift/client.py

```diff
@@ -0,0 +1,258 @@
+                ns=namespace
+            ) if apply_template else ""
+        )
+        return self.run_command(cmd=command, shell=True)
+
+    @retry(tries=10, delay=3, backoff=2)
+    def start_build(self, namespace, bc):
+        """
+        Starts build on a BuildConfig with build_name in specified
+        namespace
+        :param namespace: The namespace in which the build is to be
+        started
+        :param bc: The name of the BuildConfig
+        :type namespace str
+        :type bc str
+        :return: The output of the executed command
+        :rtype str
+        :raises Exception
+        """
+        command = "{base_cmd} start-build {build_name} -n {namespace}".format(
+            base_cmd=self.base_command, build_name=bc, namespace=namespace
+        )
+        return self.run_command(cmd=command, shell=False)
+
+    @retry(tries=10, delay=3, backoff=2)
+    def delete_build_config(self, namespace, bc):
+        """
+        Deletes a specified BuildConfig in a specified namespace
+        :param namespace: The namespcae from which the BuildConfig is to be
+        removed
+        :param bc: The name of the BuildConfig to delete
+        :type namespace str
+        :type bc str
+        :return: output of executed command
+        :rtype str
+        :raises Exception
+        """
+        command = "{base_command} delete -n {namespace} bc {bc}".format(
+            base_command=self.base_command,
+            namespace=namespace,
+            bc=bc,
+        )
+        return self.run_command(cmd=command, shell=False)
+
+    @retry(tries=10, delay=3, backoff=2)
+    def list_build_configs(self, namespace, filter_out=None, selectors=None):
+        """
+        Lists all the BuildConfigs in specified namespace except those
+        requested to be filtered out.
+        :param namespace: The namespace from which to get the build configs.
+        :param filter_out: Default None: A list of BuildConfig names to filter
+        out.
+        :param selectors Filter by selectors, provides list of labels
+        :type namespace str
+        :type filter_out list
+        :type selectors dict
+        :return: A list of BuildConfigs in namespace, after filtering
+        :rtype list
+        :raises Exception
+        """
+        if not filter_out or not isinstance(filter_out, list):
+            filter_out = []
+        selector_params = ""
+        for k, v in selectors:
+            selector_params = selector_params + " --selector {k}={v}".format(
+                k=k,
+                v=v
+            )
+        command = "{base_command} get bc -o name -n {namespace}{s}".format(
+            base_command=self.base_command,
+            namespace=namespace,
+            s=selector_params
+        )
+        out = self.run_command(cmd=command)
+        if not out.strip():
+            return []
+        out = out.strip().split('\n')
+        return [each for each in out if each not in filter_out]
```

Comment:
got puzzled here, this returns a list or string? also can we use some variable name instead of `each`?

Reply:
Yes.  This is list comprehension. It will return a list of items.  This is basically form a list from out.  For each item in out.  If that item is not in filter_out which is another list.  It will be added to the list that will be returned

Reply:
If u remember mathematical notations from set theory, this is equivalent to
{each:each € out and each € with / filter out }